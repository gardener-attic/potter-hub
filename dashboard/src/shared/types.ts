import * as jsonSchema from "json-schema";
import { IAuthState } from "../reducers/auth";
import { IClusterBomState } from "../reducers/clusterbom";
import { IConfigState } from "../reducers/config";
import { INamespaceState } from "../reducers/namespace";
import { IAppRepositoryState } from "../reducers/repos";
import { IServiceCatalogState } from "../reducers/serviceCatalog";
import { hapi } from "./hapi/release";
import { ResourceKind } from "./ResourceKinds";

// Allow defining multiple error classes
// tslint:disable:max-classes-per-file

class CustomError extends Error {
  // The constructor is defined so we can later on compare the returned object
  // via err.contructor  == FOO
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
export class ForbiddenError extends CustomError {}
export class UnauthorizedError extends CustomError {}

export class NotFoundError extends CustomError {}

export class ConflictError extends CustomError {}

export class UnprocessableEntity extends CustomError {}

export interface IRepo {
  name: string;
  url: string;
}

export interface IChartVersion {
  id: string;
  attributes: IChartVersionAttributes;
  relationships: {
    chart: {
      data: IChartAttributes;
    };
  };
}

export interface IChartVersionAttributes {
  version: string;
  app_version: string;
  created: string;
}

export interface IChart {
  id: string;
  attributes: IChartAttributes;
  relationships: {
    latestChartVersion: {
      data: IChartVersionAttributes;
    };
  };
}

export interface IChartAttributes {
  name: string;
  description: string;
  home?: string;
  icon?: string;
  keywords: string[];
  maintainers: Array<{
    name: string;
    email?: string;
  }>;
  repo: IRepo;
  sources: string[];
}

export interface IChartState {
  isFetching: boolean;
  selected: {
    error?: Error;
    version?: IChartVersion;
    versions: IChartVersion[];
    readme?: string;
    readmeError?: string;
    values?: string;
    schema?: any;
  };
  deployed: {
    chartVersion?: IChartVersion;
    values?: string;
    schema?: jsonSchema.JSONSchema4;
  };
  items: IChart[];
}

export interface IChartUpdateInfo {
  upToDate: boolean;
  chartLatestVersion: string;
  appLatestVersion: string;
  repository: IRepo;
  error?: Error;
}

export interface IDeployment {
  metadata: {
    name: string;
    namespace: string;
  };
}

export interface IServiceSpec {
  ports: IPort[];
  clusterIP: string;
  type: string;
}

export interface IServiceStatus {
  loadBalancer: {
    ingress?: Array<{ ip?: string; hostname?: string }>;
  };
}

export interface IPort {
  name: string;
  port: number;
  protocol: string;
  targetPort: string;
  nodePort: string;
}

export interface IHTTPIngressPath {
  path: string;
}
export interface IIngressHTTP {
  paths: IHTTPIngressPath[];
}
export interface IIngressRule {
  host: string;
  http: IIngressHTTP;
}

export interface IIngressTLS {
  hosts: string[];
}

export interface IIngressSpec {
  rules: IIngressRule[];
  tls?: IIngressTLS[];
}

export interface IResource {
  apiVersion: string;
  kind: ResourceKind;
  type: string;
  spec: any;
  status: any;
  metadata: {
    name: string;
    namespace: string;
    annotations?: { [key: string]: string };
    creationTimestamp: string;
    deletionTimestamp?: string | null;
    generation?: number;
    labels?: { [key: string]: string };
    resourceVersion: string;
    uid: string;
    selfLink: string;
  };
}

export interface IOwnerReference {
  apiVersion: string;
  blockOwnerDeletion: boolean;
  kind: string;
  name: string;
  uid: string;
}

export interface ISecret {
  apiVersion: string;
  kind: string;
  type: string;
  data: { [s: string]: string };
  metadata: {
    name: string;
    namespace: string;
    annotations: string;
    creationTimestamp: string;
    selfLink: string;
    resourceVersion: string;
    deletionTimestamp?: string;
    uid: string;
  };
}

export interface IDeploymentStatus {
  replicas: number;
  updatedReplicas: number;
  availableReplicas: number;
}

export interface IStatefulsetStatus {
  replicas: number;
  updatedReplicas: number;
  readyReplicas: number;
}

export interface IDaemonsetStatus {
  currentNumberScheduled: number;
  numberReady: number;
}

export interface IRelease extends hapi.release.Release {
  updateInfo?: IChartUpdateInfo;
}

export interface IAppState {
  isFetching: boolean;
  error?: Error;
  deleteError?: Error;
  // currently items are always Helm releases
  items: IRelease[];
  listingAll: boolean;
  listOverview?: IAppOverview[];
  selected?: IRelease;
}

export interface IStoreState {
  serviceCatalog: IServiceCatalogState;
  apps: IAppState;
  auth: IAuthState;
  charts: IChartState;
  clusterBom: IClusterBomState;
  config: IConfigState;
  kube: IKubeState;
  repos: IAppRepositoryState;
  namespace: INamespaceState;
}

interface IK8sResource {
  apiVersion: string;
  kind: string;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#objects */
export interface IK8sObject<M, SP, ST> extends IK8sResource {
  metadata: {
    annotations?: { [key: string]: string };
    creationTimestamp?: string;
    deletionTimestamp?: string | null;
    generation?: number;
    labels?: { [key: string]: string };
    name: string;
    namespace: string;
    resourceVersion?: string;
    uid: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
  spec?: SP;
  status?: ST;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#lists-and-simple-kinds */
export interface IK8sList<I, M> extends IK8sResource {
  items: I[];
  metadata?: {
    resourceVersion?: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
}

export interface IAppRepository
  extends IK8sObject<
    {
      clusterName: string;
      creationTimestamp: string;
      deletionGracePeriodSeconds: string | null;
      deletionTimestamp: string | null;
      resourceVersion: string;
      selfLink: string;
    },
    {
      type: string;
      url: string;
      auth: {
        header: {
          secretKeyRef: {
            name: string;
            key: string;
          };
        };
      };
      resyncRequests: number;
    },
    undefined
  > {}

export interface IAppRepositoryList
  extends IK8sList<
    IAppRepository,
    {
      continue: string;
      resourceVersion: string;
      selfLink: string;
    }
  > {}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#response-status-kind */
export interface IStatus extends IK8sResource {
  kind: "Status";
  status: "Success" | "Failure";
  message: string;
  reason:
    | "BadRequest"
    | "Unauthorized"
    | "Forbidden"
    | "NotFound"
    | "AlreadyExists"
    | "Conflict"
    | "Invalid"
    | "Timeout"
    | "ServerTimeout"
    | "MethodNotAllowed"
    | "InternalError";
  details?: {
    kind?: string;
    name?: string;
    causes?: IStatusCause[] | string;
  };
}

interface IStatusCause {
  field: string;
  message: string;
  reason: string;
}

export interface IRouterPathname {
  router: {
    location: {
      pathname: string;
    };
  };
}

export interface IRuntimeVersion {
  name: string;
  version: string;
  runtimeImage: string;
  initImage: string;
}

export interface IRuntime {
  ID: string;
  versions: IRuntimeVersion[];
  depName: string;
  fileNameSuffix: string;
}

export interface IRBACRole {
  apiGroup: string;
  namespace?: string;
  clusterWide?: boolean;
  resource: string;
  verbs: string[];
}

export interface IAppOverview {
  releaseName: string;
  namespace: string;
  version: string;
  icon?: string;
  status: string;
  chart: string;
  chartMetadata: hapi.chart.Metadata;
  // UpdateInfo is internally populated
  updateInfo?: IChartUpdateInfo;
  description: string;
}

export interface IKubeItem<T> {
  isFetching: boolean;
  item?: T;
  error?: Error;
}

export interface IKubeState {
  items: { [s: string]: IKubeItem<IResource> };
  fetchTargetClusterSecretErr?: Error;
  sockets: { [s: string]: { socket: WebSocket; closeTimer: () => void } };
}

export interface IBasicFormParam {
  path: string;
  type?: jsonSchema.JSONSchema4TypeName | jsonSchema.JSONSchema4TypeName[];
  value?: any;
  title?: string;
  minimum?: number;
  maximum?: number;
  render?: string;
  description?: string;
  enum?: string[];
  hidden?:
    | {
        path: any;
        value: string;
        conditions: Array<{
          path: any;
          value: string;
        }>;
        operator: string;
      }
    | string;
  children?: IBasicFormParam[];
}
export interface IBasicFormSliderParam extends IBasicFormParam {
  sliderMin?: number;
  sliderMax?: number;
  sliderUnit?: string;
}

export interface IApplicationConfig {
  id: string;
  configType: string;
  typeSpecificData: {};
  values: {};
  secretValues: {};
  noReconcile: boolean;
  readyRequirements: IReadyRequirements;
}

export interface IReadyRequirements {
    jobs: IJobReadyRequirement[];
    resources: IResourceReadyRequirement[];
}

export interface IJobReadyRequirement {
    name: string;
    namespace: string
}

export interface IResourceReadyRequirement {
    name: string;
    namespace: string;
    apiVersion: string;
    resource: string;
    fieldPath: string;
    successValues: ISuccessValue[];
}

export interface ISuccessValue {
    value: {};
}

export interface IAutoDelete {
    clusterBomAge: number;
}

export interface IClusterBom
  extends IK8sObject<
    {},
    {
      secretRef: string;
      applicationConfigs: IApplicationConfig[];
      autoDelete: IAutoDelete;
    },
    {
      observedGeneration: number;
      overallState: string;
      overallTime: string;
      conditions: ICondition[];
      applicationStates: IApplicationState[];
      overallNumOfDeployments: number;
      overallNumOfReadyDeployments: number;
      overallProgress: number;
    }
  > {}

export interface ICondition {
  type: string;
  status: string;
  lastUpdateTime: string;
  lastTransitionTime: string;
  reason: string;
  message: string;
}

export interface IApplicationState {
  id: string;
  state: string;
  detailedState: {};
}

export interface IClusterBomList extends IK8sList<IClusterBom, {}> {}

export interface IURLParams {
    targetClusterSecretName: string;
    targetClusterSecretNamespace: string;
}
