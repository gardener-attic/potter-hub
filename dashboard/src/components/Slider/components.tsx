// CODE FROM: https://sghall.github.io/react-compound-slider/#/slider-demos/horizontal
import * as React from "react";
import { GetHandleProps, GetTrackProps, SliderItem } from "react-compound-slider";

// *******************************************************
// HANDLE COMPONENT
// *******************************************************
interface IHandleProps {
  domain: number[];
  handle: SliderItem;
  getHandleProps: GetHandleProps;
}

export const Handle: React.SFC<IHandleProps> = ({
  domain: [min, max],
  handle: { id, value, percent },
  getHandleProps,
}) => (
  <div
    role="slider"
    aria-valuemin={min}
    aria-valuemax={max}
    aria-valuenow={value}
    style={{
      left: `${percent}%`,
      position: "absolute",
      marginLeft: "-11px",
      marginTop: "-6px",
      zIndex: 2,
      width: 24,
      height: 24,
      cursor: "pointer",
      borderRadius: "50%",
      boxShadow: "1px 1px 1px 1px rgba(0, 0, 0, 0.2)",
      backgroundColor: "#34568f",
    }}
    {...getHandleProps(id)}
  />
);

// *******************************************************
// TRACK COMPONENT
// *******************************************************
interface ITrackProps {
  source: SliderItem;
  target: SliderItem;
  getTrackProps: GetTrackProps;
}

export const Track: React.SFC<ITrackProps> = ({ source, target, getTrackProps }) => (
  <div
    style={{
      position: "absolute",
      height: 14,
      zIndex: 1,
      backgroundColor: "#7aa0c4",
      borderRadius: 7,
      cursor: "pointer",
      left: `${source.percent}%`,
      width: `${target.percent - source.percent}%`,
    }}
    {...getTrackProps()}
  />
);
