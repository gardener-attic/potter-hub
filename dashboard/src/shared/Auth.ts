import Axios, { AxiosResponse } from "axios";
import * as jwt from "jsonwebtoken";
import { K8S_REVERSE_PROXY_URL } from "./Kube";

const AuthTokenKey = "kubeapps_auth_token";
const AuthTokenOIDCKey = "kubeapps_auth_token_oidc";

export const DEFAULT_NAMESPACE = "_all";

export class Auth {
  public static getAuthToken() {
    return localStorage.getItem(AuthTokenKey);
  }

  public static setAuthToken(token: string, oidc: boolean) {
    localStorage.setItem(AuthTokenOIDCKey, oidc.toString());
    return localStorage.setItem(AuthTokenKey, token);
  }

  public static unsetAuthToken() {
    return localStorage.removeItem(AuthTokenKey);
  }

  public static usingOIDCToken() {
    return localStorage.getItem(AuthTokenOIDCKey) === "true";
  }

  public static wsProtocols() {
    const token = this.getAuthToken();
    if (!token) {
      return [];
    }
    return ["base64url.bearer.authorization.k8s.io." + token, "binary.k8s.io"];
  }

  public static fetchOptions(): RequestInit {
    const headers = new Headers();
    headers.append("Authorization", `Bearer ${this.getAuthToken()}`);
    return {
      headers,
    };
  }

  // Throws an error if the token is invalid
  public static async validateToken(token: string) {
    try {
      await Axios.get(K8S_REVERSE_PROXY_URL + "/", {
        headers: { Authorization: `Bearer ${token}` },
      });
    } catch (e) {
      const res = e.response as AxiosResponse;
      if (res.status === 401) {
        throw new Error("invalid token");
      } else if (res.status === 403) {
        throw new Error("token not authorised for cluster");
      }
      // A 403 authorization error only occurs if the token resulted in
      // successful authentication. We don't make any assumptions over RBAC
      // for the root "/" nonResourceURL or other required authz permissions
      // until operations on those resources are attempted (though we may
      // want to revisit this in the future).
      // In the hub project we want to reject everyone with a 403 error
      // since he can't do anything in the ui anyway. We only have
      // admin or unauthorised users.
      if (res.status !== 403) {
        throw new Error(`${res.status}: ${res.data}`);
      }
    }
  }

  // fetchOIDCToken does a HEAD request to collect the Bearer token
  // from the authorization header if exists
  public static async fetchOIDCToken(): Promise<string | null> {
    try {
      const { headers } = await Axios.head("");
      if (headers && headers.authorization) {
        const tokenMatch = (headers.authorization as string).match(/Bearer\s(.*)/);
        if (tokenMatch) {
          return tokenMatch[1];
        }
      }
    } catch (e) {
      // Unable to retrieve token
    }
    return null;
  }

  // defaultNamespaceFromToken decodes a jwt token to return the k8s service
  // account namespace.
  // TODO(mnelson): until we call jwt.verify on the token during validateToken above
  // we use a default namespace for both invalid tokens and tokens without the expected
  // key.
  public static defaultNamespaceFromToken(token: string) {
    const payload = jwt.decode(token);
    const namespaceKey = "kubernetes.io/serviceaccount/namespace";
    if (!payload || !payload[namespaceKey]) {
      return DEFAULT_NAMESPACE;
    }
    return payload[namespaceKey];
  }
}
