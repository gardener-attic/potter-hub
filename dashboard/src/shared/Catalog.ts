export const allRepositories = {
  value: "_all",
  label: "All repositories",
};
