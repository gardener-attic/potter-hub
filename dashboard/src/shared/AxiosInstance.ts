import Axios, { AxiosError, AxiosInstance, AxiosRequestConfig } from "axios";

import { Action, Store } from "redux";
import { ThunkDispatch } from "redux-thunk";
import actions from "../actions";
import { Auth } from "./Auth";
import {
  ConflictError,
  ForbiddenError,
  IStoreState,
  NotFoundError,
  UnauthorizedError,
  UnprocessableEntity,
} from "./types";

export function addAuthHeaders(axiosInstance: AxiosInstance) {
  axiosInstance.interceptors.request.use((config: AxiosRequestConfig) => {
    const authToken = Auth.getAuthToken();
    if (authToken) {
      config.headers.Authorization = `Bearer ${authToken}`;
    }
    return config;
  });
}

export function addErrorHandling(axiosInstance: AxiosInstance, store: Store<IStoreState>) {
  axiosInstance.interceptors.response.use(
    response => response,
    e => {
      const dispatch = store.dispatch as ThunkDispatch<IStoreState, null, Action>;
      const err: AxiosError = e;
      if (
        err.code === undefined &&
        err.message === "Network Error" &&
        !err.response &&
        Auth.usingOIDCToken()
      ) {
        // The OIDC token is no longer valid, logout
        dispatch(actions.auth.expireSession());
      }
      let message = err.message;
      if (err.response && err.response.data.message) {
        message = err.response.data.message;
      }
      switch (err.response && err.response.status) {
        case 401:
          // Global action dispatch to log the user out
          // Expire the session if we are using OIDC tokens
          dispatch(actions.auth.authenticationError(message));
          dispatch(actions.auth.expireSession());
          return Promise.reject(new UnauthorizedError(message));
        case 403:
          return Promise.reject(new ForbiddenError(message));
        case 404:
          return Promise.reject(new NotFoundError(message));
        case 409:
          return Promise.reject(new ConflictError(message));
        case 422:
          return Promise.reject(new UnprocessableEntity(message));
        default:
          return Promise.reject(new Error(message));
      }
    },
  );
}

// Error handling is added with an interceptor in index.tsx
export const axios = Axios.create();
// Authorization headers and error handling are added with an interceptor in index.tsx
const a = Axios.create();
a.defaults.timeout = 1800000;

export const axiosWithAuth = a;
