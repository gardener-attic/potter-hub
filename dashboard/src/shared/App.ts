import { axiosWithAuth } from "./AxiosInstance";
import { hapi } from "./hapi/release";
import { HELM_URL } from "./Kube";
import { IAppOverview, IChartVersion } from "./types";

export class App {
  public static getResourceURL(namespace?: string, name?: string, query?: string) {
    let url = HELM_URL;
    if (namespace) {
      url += `/namespaces/${namespace}`;
    }
    url += "/releases";
    if (name) {
      url += `/${name}`;
    }
    if (query) {
      url += `?${query}`;
    }
    return url;
  }

  public static async create(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = App.getResourceURL(namespace);
    const { data } = await axiosWithAuth.post(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async upgrade(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = App.getResourceURL(namespace, releaseName);
    const { data } = await axiosWithAuth.put(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async rollback(releaseName: string, namespace: string, revision: number) {
    const endpoint = App.getResourceURL(namespace, releaseName);
    const { data } = await axiosWithAuth.put(
      endpoint,
      {},
      {
        params: {
          action: "rollback",
          revision,
        },
      },
    );
    return data;
  }

  public static async delete(releaseName: string, namespace: string, keepHistory: boolean) {
    let keepHistoryQuery;
    if (keepHistory) {
      keepHistoryQuery = "keepHistory=true";
    }
    const { data } = await axiosWithAuth.delete(
      App.getResourceURL(namespace, releaseName, keepHistoryQuery),
    );
    return data;
  }

  public static async listApps(namespace?: string, allStatuses?: boolean) {
    let query;
    if (allStatuses) {
      query = "statuses=all";
    }
    const { data } = await axiosWithAuth.get<{ data: IAppOverview[] }>(
      App.getResourceURL(namespace, undefined, query),
    );
    return data.data;
  }

  public static async getRelease(namespace: string, name: string) {
    const { data } = await axiosWithAuth.get<{ data: hapi.release.Release }>(
      this.getResourceURL(namespace, name),
    );
    return data.data;
  }
}
