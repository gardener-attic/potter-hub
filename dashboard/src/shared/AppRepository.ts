import { axiosWithAuth } from "./AxiosInstance";
import { UI_BACKEND_ROOT_URL } from "./Kube";
import { IAppRepository, IAppRepositoryList } from "./types";

export class AppRepository {
  public static async list(namespace: string) {
    const { data } = await axiosWithAuth.get<IAppRepositoryList>(
      AppRepository.getResourceLink(namespace),
    );
    return data;
  }

  public static async get(name: string, namespace: string) {
    const { data } = await axiosWithAuth.get(AppRepository.getSelfLink(name, namespace));
    return data;
  }

  public static async update(name: string, namespace: string, newApp: IAppRepository) {
    const { data } = await axiosWithAuth.put(AppRepository.getSelfLink(name, namespace), newApp);
    return data;
  }

  public static async delete(name: string, namespace: string) {
    const { data } = await axiosWithAuth.delete(AppRepository.getSelfLink(name, namespace));
    return data;
  }

  public static async create(
    name: string,
    namespace: string,
    url: string,
    auth: any,
    syncJobPodTemplate: any,
  ) {
    const { data } = await axiosWithAuth.post<IAppRepository>(
      AppRepository.getResourceLink(namespace),
      {
        apiVersion: "kubeapps.com/v1alpha1",
        kind: "AppRepository",
        metadata: {
          name,
        },
        spec: { auth, type: "helm", url, syncJobPodTemplate },
      },
    );
    return data;
  }

  private static APIBase: string = UI_BACKEND_ROOT_URL;
  private static APIEndpoint: string = `${AppRepository.APIBase}/apprepositories`;
  
  private static getResourceLink(namespace?: string): string {
    return AppRepository.APIEndpoint;
  }

  private static getSelfLink(name: string, namespace: string): string {
    return `${AppRepository.APIEndpoint}/${name}`;
  }
}
