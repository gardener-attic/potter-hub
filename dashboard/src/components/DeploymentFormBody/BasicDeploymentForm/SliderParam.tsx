import * as React from "react";
import { IBasicFormParam } from "shared/types";
import Slider from "../../Slider";

export interface ISliderParamProps {
  id: string;
  label: string;
  param: IBasicFormParam;
  unit: string;
  min: number;
  max: number;
  handleBasicFormParamChange: (
    p: IBasicFormParam,
  ) => (e: React.FormEvent<HTMLInputElement>) => void;
}

export interface ISliderParamState {
  value: number;
}

function toNumber(value: string) {
  // Force to return a Number from a string removing any character that is not a digit
  return Number(value.replace(/[^\d\.]/g, ""));
}

function getDefaultValue(min: number, value?: string) {
  return (value && toNumber(value)) || min;
}

class SliderParam extends React.Component<ISliderParamProps, ISliderParamState> {
  public state: ISliderParamState = {
    value: getDefaultValue(this.props.min, this.props.param.value),
  };

  public componentDidUpdate = (prevProps: ISliderParamProps) => {
    if (prevProps.param.value !== this.props.param.value) {
      const value = getDefaultValue(this.props.min, this.props.param.value);
      if (value !== this.state.value) {
        this.setState({ value });
      }
    }
  };

  // onChangeSlider is executed when the slider is dropped at one point
  // at that point we update the parameter
  public onChangeSlider = (values: readonly number[]) => {
    this.handleParamChange(values[0]);
  };

  // onUpdateSlider is executed when dragging the slider
  // we just update the state here for a faster response
  public onUpdateSlider = (values: readonly number[]) => {
    this.setState({ value: values[0] });
  };

  public onChangeInput = (e: React.FormEvent<HTMLInputElement>) => {
    const value = toNumber(e.currentTarget.value);
    this.setState({ value });
    this.handleParamChange(value);
  };

  public render() {
    const { param, label, min, max } = this.props;
    return (
      <div>
        <label htmlFor={this.props.id}>
          <div className="row">
            <div className="col-3 block">
              <div className="centered">{label}</div>
            </div>
            <div className="col-9 margin-t-small">
              <div className="row">
                <div className="col-9">
                  <Slider
                    // If the parameter defines a minimum or maximum, maintain those
                    min={Math.min(param.minimum || min, this.state.value)}
                    max={Math.max(param.maximum || max, this.state.value)}
                    default={this.state.value}
                    onChange={this.onChangeSlider}
                    onUpdate={this.onUpdateSlider}
                    values={this.state.value}
                  />
                </div>
                <div className="col-3">
                  <input
                    className="disk_size_input"
                    id={this.props.id}
                    onChange={this.onChangeInput}
                    value={this.state.value}
                  />
                  <span className="margin-l-normal">{this.props.unit}</span>
                </div>
              </div>
              {param.description && <span className="description">{param.description}</span>}
            </div>
          </div>
        </label>
      </div>
    );
  }

  private handleParamChange = (value: number) => {
    this.props.handleBasicFormParamChange(this.props.param)({
      currentTarget: { value: `${value}${this.props.unit}` },
    } as React.FormEvent<HTMLInputElement>);
  };
}

export default SliderParam;
