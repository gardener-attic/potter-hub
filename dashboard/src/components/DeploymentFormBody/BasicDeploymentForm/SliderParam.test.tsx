import { shallow } from "enzyme";
import * as React from "react";
import { IBasicFormParam } from "shared/types";
import Slider from "../../Slider";
import SliderParam from "./SliderParam";

const defaultProps = {
  id: "disk",
  label: "Disk Size",
  param: {
    value: "10Gi",
    type: "string",
    path: "disk",
  } as IBasicFormParam,
  handleBasicFormParamChange: jest.fn(() => jest.fn()),
  min: 1,
  max: 100,
  unit: "Gi",
};

it("renders a disk size param with a default value", () => {
  const wrapper = shallow(<SliderParam {...defaultProps} />);
  expect(wrapper.state("value")).toBe(10);
  expect(wrapper).toMatchSnapshot();
});

it("changes the value of the param when the slider changes", () => {
  const param = {
    value: "10Gi",
    type: "string",
    path: "disk",
  } as IBasicFormParam;
  const handleBasicFormParamChange = jest.fn(() => {
    param.value = "20Gi";
    return jest.fn();
  });

  const wrapper = shallow(
    <SliderParam
      {...defaultProps}
      param={param}
      handleBasicFormParamChange={handleBasicFormParamChange}
    />,
  );
  expect(wrapper.state("value")).toBe(10);

  const slider = wrapper.find(Slider);
  (slider.prop("onChange") as (values: number[]) => void)([20]);

  expect(param.value).toBe("20Gi");
  expect(handleBasicFormParamChange.mock.calls[0]).toEqual([
    { value: "20Gi", type: "string", path: "disk" },
  ]);
});

it("updates state but does not change param value during slider update (only when dropped in a point)", () => {
  const handleBasicFormParamChange = jest.fn();
  const wrapper = shallow(
    <SliderParam {...defaultProps} handleBasicFormParamChange={handleBasicFormParamChange} />,
  );
  expect(wrapper.state("value")).toBe(10);

  const slider = wrapper.find(Slider);
  (slider.prop("onUpdate") as (values: number[]) => void)([20]);

  expect(wrapper.state("value")).toBe(20);
  expect(handleBasicFormParamChange).not.toHaveBeenCalled();
});

describe("when changing the value in the input", () => {
  it("parses a number and forwards it", () => {
    const valueChange = jest.fn();
    const handleBasicFormParamChange = jest.fn(() => valueChange);
    const wrapper = shallow(
      <SliderParam {...defaultProps} handleBasicFormParamChange={handleBasicFormParamChange} />,
    );
    expect(wrapper.state("value")).toBe(10);

    const input = wrapper.find("input#disk");
    const event = { currentTarget: { value: "20" } } as React.FormEvent<HTMLInputElement>;
    (input.prop("onChange") as (e: React.FormEvent<HTMLInputElement>) => void)(event);

    expect(wrapper.state("value")).toBe(20);
    expect(valueChange.mock.calls[0]).toEqual([{ currentTarget: { value: "20Gi" } }]);
  });

  it("ignores values in the input that are not digits", () => {
    const valueChange = jest.fn();
    const handleBasicFormParamChange = jest.fn(() => valueChange);
    const wrapper = shallow(
      <SliderParam {...defaultProps} handleBasicFormParamChange={handleBasicFormParamChange} />,
    );
    expect(wrapper.state("value")).toBe(10);

    const input = wrapper.find("input#disk");
    const event = { currentTarget: { value: "foo20*#@$" } } as React.FormEvent<HTMLInputElement>;
    (input.prop("onChange") as (e: React.FormEvent<HTMLInputElement>) => void)(event);

    expect(wrapper.state("value")).toBe(20);
    expect(valueChange.mock.calls[0]).toEqual([{ currentTarget: { value: "20Gi" } }]);
  });

  it("accept decimal values", () => {
    const valueChange = jest.fn();
    const handleBasicFormParamChange = jest.fn(() => valueChange);
    const wrapper = shallow(
      <SliderParam {...defaultProps} handleBasicFormParamChange={handleBasicFormParamChange} />,
    );
    expect(wrapper.state("value")).toBe(10);

    const input = wrapper.find("input#disk");
    const event = { currentTarget: { value: "20.5" } } as React.FormEvent<HTMLInputElement>;
    (input.prop("onChange") as (e: React.FormEvent<HTMLInputElement>) => void)(event);

    expect(wrapper.state("value")).toBe(20.5);
    expect(valueChange.mock.calls[0]).toEqual([{ currentTarget: { value: "20.5Gi" } }]);
  });

  it("modifies the max value of the slider if the input is bigger than 100", () => {
    const valueChange = jest.fn();
    const handleBasicFormParamChange = jest.fn(() => valueChange);
    const wrapper = shallow(
      <SliderParam {...defaultProps} handleBasicFormParamChange={handleBasicFormParamChange} />,
    );
    expect(wrapper.state("value")).toBe(10);

    const input = wrapper.find("input#disk");
    const event = { currentTarget: { value: "200" } } as React.FormEvent<HTMLInputElement>;
    (input.prop("onChange") as (e: React.FormEvent<HTMLInputElement>) => void)(event);

    expect(wrapper.state("value")).toBe(200);
    const slider = wrapper.find(Slider);
    expect(slider.prop("max")).toBe(200);
  });
});

it("uses the param minimum and maximum if defined", () => {
  const param = {
    value: "10Gi",
    type: "string",
    path: "disk",
    minimum: 5,
    maximum: 50,
  } as IBasicFormParam;

  const wrapper = shallow(<SliderParam {...defaultProps} param={param} />);

  const slider = wrapper.find(Slider);
  expect(slider.prop("min")).toBe(5);
  expect(slider.prop("max")).toBe(50);
});

it("defaults to the min if the value is undefined", () => {
  const param = {
    type: "string",
    path: "disk",
  } as IBasicFormParam;

  const wrapper = shallow(<SliderParam {...defaultProps} param={param} min={5} />);

  expect(wrapper.state("value")).toBe(5);
});

it("updates the state when receiving new props", () => {
  const handleBasicFormParamChange = jest.fn();
  const wrapper = shallow(
    <SliderParam {...defaultProps} handleBasicFormParamChange={handleBasicFormParamChange} />,
  );
  expect(wrapper.state("value")).toBe(10);

  wrapper.setProps({ param: { value: "20Gi" } });
  expect(wrapper.state("value")).toBe(20);
  expect(handleBasicFormParamChange).not.toHaveBeenCalled();
});
