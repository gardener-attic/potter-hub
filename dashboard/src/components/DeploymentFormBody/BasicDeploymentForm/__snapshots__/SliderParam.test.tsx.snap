// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a disk size param with a default value 1`] = `
<div>
  <label
    htmlFor="disk"
  >
    <div
      className="row"
    >
      <div
        className="col-3 block"
      >
        <div
          className="centered"
        >
          Disk Size
        </div>
      </div>
      <div
        className="col-9 margin-t-small"
      >
        <div
          className="row"
        >
          <div
            className="col-9"
          >
            <Slider
              default={10}
              max={100}
              min={1}
              onChange={[Function]}
              onUpdate={[Function]}
              values={10}
            />
          </div>
          <div
            className="col-3"
          >
            <input
              className="disk_size_input"
              id="disk"
              onChange={[Function]}
              value={10}
            />
            <span
              className="margin-l-normal"
            >
              Gi
            </span>
          </div>
        </div>
      </div>
    </div>
  </label>
</div>
`;
